### PR TITLE
OTA-4594: add a job to the CI to build aktualizr on OSX

### PR DESCRIPTION
**Pull request by mike-sul**
The least intrusive way to add the brew build job into the aktualizr CI pipeline.
The job is included in the pipeline only if executed as 'scheduled` at `docker` stage.

Signed-off-by: Mike Sul <ext-mykhaylo.sul@here.com>

### CODE REVIEW

File: ci/gitlab/.gitlab-ci.yml [hunk 1]
```diff
@@ -408,3 +408,18 @@ trigger-docsite-build:
...
+    - brew uninstall -f aktualizr-dev.rb
+  script:
+    - brew tap advancedtelematic/otaconnect
+    - brew install --HEAD -v aktualizr-dev
+    - brew link -f aktualizr-dev
+    - aktualizr --version
+  tags:
+    - osx
+  only:
+    - schedules
```
Comment (lbonn) [inline ci/gitlab/.gitlab-ci.yml]: We have another weekly schedule for Veracode scans, we probably don't need to run the osx-build in that case.

Using `only: variables: - $OSX_BUILD` and defining the variable in the schedule would solve that.

Also, it makes it easier to trigger the job manually without having to go through a schedule.

Comment (mike-sul) [reply ci/gitlab/.gitlab-ci.yml]: The scheduling mechanism allows triggering a job manually too.
Anyway, I agree with you, it makes sense

**Comment by mike-sul** [comment]
BTW, if I trigger the pipeline from the schedules https://main.gitlab.in.here.com/olp/edge/ota/connect/client/aktualizr/pipeline_schedules that pipeline run/instance is attached to PR :)

**Comment by patrickvacek** [comment]
Looks cool!

**Comment by mike-sul** [comment]
@patrickvacek @lbonn Please, see this option of implementation - triggering an external project CI pipeline (https://main.gitlab.in.here.com/olp/edge/ota/connect/client/aktualizr/pipelines/402012)

In addition to triggering of the aktualizr' brewing job/pipeline I could also add triggering of the aktualizr release version brewing and adding it to the release artifacts if `github-release` is triggered

**Comment by patrickvacek** [comment]
> Please, see this option of implementation - triggering an external project CI pipeline

Looks good!

> In addition to triggering of the aktualizr' brewing job/pipeline I could also add triggering of the aktualizr release version brewing and adding it to the release artifacts if `github-release` is triggered

That would be the actual replacement for the github action that we were using to do that, right? Seems like the right thing to do.

**Comment by mike-sul** [comment]
> That would be the actual replacement for the github action that we were using to do that, right? Seems like the right thing to do.

Replacement or addition if we would like to support both mojave (the mac in the lab) and catalina (macOS virtual machine provided by github)?

**Comment by patrickvacek** [comment]
> Replacement or addition if we would like to support both mojave (the mac in the lab) and catalina (macOS virtual machine provided by github)?

I think @tkfu suggested that it wasn't worth maintaining the github action, and I'm inclined to agree.

**Comment by mike-sul** [comment]
> I think @tkfu suggested that it wasn't worth maintaining the github action, and I'm inclined to agree.

OK. Agreed.

**Comment by patrickvacek** [comment]
> It's still itching a bit to see the step under the docker stage but we can think about another organization later.

What's your concern with that?

@mike-sul this has to wait until the homebrew recipe is merged so that you can update the branchname that's hardcoded here, right?

**Comment by mike-sul** [comment]
> @mike-sul this has to wait until the homebrew PR is merged so that you can update the branchname that's hardcoded here, right?

Yes, that's correct + I would like to add also the job to build the release version as well as trying to improve the existing two formulae so we have just one instead of two.

**Comment by mike-sul** [comment]
> It's still itching a bit to see the step under the docker stage but we can think about another organization later.

I don't like it too, but the reorganization of the CI stages will take some time and it's not primary goal of this specific task/PR (Frankly speaking I expected to see something like build, test, deploy, end-to-end test, etc)

**Comment by lbonn** [comment]
> > It's still itching a bit to see the step under the docker stage but we can think about another organization later.
> 
> What's your concern with that?

It has nothing to do with docker, so it's confusing. We could add a comment in the meantime, or a new stage ("early-trigger"?) before "docker".

**Comment by mike-sul** [comment]
> It has nothing to do with docker, so it's confusing. We could add a comment in the meantime, or a new stage ("early-trigger"?) before "docker".

or it can be part of `test` stage?

**Comment by lbonn** [comment]
@mike-sul I thought the reason you put it in docker was to execute it as soon as possible. If it goes in "test", then why not "trigger" which fires up at the end?

**Comment by mike-sul** [comment]
> @mike-sul I thought the reason you put it in docker was to execute it as soon as possible. If it goes in "test", then why not "trigger" which fires up at the end?

Yes, that was my intention to trigger it ASAP, but the rationale behind it just to see results ASAP while I am working on it. I open for suggestions regarding a stage at which the job should be triggered.

**Comment by patrickvacek** [comment]
> It has nothing to do with docker, so it's confusing.

Oh, I see. I misunderstood the connection there. I agree now but don't have a strong feeling about where it ends up.

**Comment by mike-sul** [comment]
I think this is ready for a merge, just made the 2020.3 OSX release with the CI job defined in this PR.
https://github.com/advancedtelematic/aktualizr/releases
https://main.gitlab.in.here.com/olp/edge/ota/connect/client/aktualizr/-/jobs/1776855

File: ci/gitlab/.gitlab-ci.yml [hunk 2]
```diff
@@ -408,3 +408,44 @@ trigger-docsite-build:
...
+    - brew install ghr
+  script:
+    - brew tap advancedtelematic/otaconnect
+    - brew install -v --build-bottle aktualizr
+    - aktualizr --version
+    - brew bottle --json --no-rebuild --force-core-tap --root-url=${RELEASE_BASE_URL}/${VERSION} aktualizr
+    - brew bottle --merge --write --no-commit ./aktualizr--${VERSION}.mojave.bottle.json
+    - brew cat aktualizr | tee aktualizr.rb
+    - mv aktualizr--${VERSION}.mojave.bottle.tar.gz aktualizr-${VERSION}.mojave.bottle.tar.gz
+    - ls -alrt aktualizr-${VERSION}.mojave.bottle.tar.gz
```
Comment (patrickvacek) [inline ci/gitlab/.gitlab-ci.yml]: `ls`?

File: ci/gitlab/.gitlab-ci.yml [hunk 3]
```diff
@@ -408,3 +408,44 @@ trigger-docsite-build:
...
     project: olp/edge/ota/documentation/ota-connect-docs
     branch: master
+
+trigger-osx-build:
+  stage: trigger
+  trigger:
+    project: olp/edge/ota/connect/client/homebrew-otaconnect
+    branch: master 
+  rules:
+    - if: $OSX_BUILD
```
Comment (patrickvacek) [inline ci/gitlab/.gitlab-ci.yml]: I didn't realize you could do this. Cool!

File: ci/gitlab/.gitlab-ci.yml (hunk 2 above)
Comment (mike-sul) [reply ci/gitlab/.gitlab-ci.yml]: just for testing/debugging purposes - it does not cause any issue while might be helpful for troubleshooting if something goes wrong.

**Comment by mike-sul** [comment]
@patrickvacek So, I am merging it?

**Pull request merged**